### DISCUSSION TITLE
error: attribute 'callCabal2Nix' missing

### DISCUSSION THREAD

**Discussion by billksun** [srid/haskell-template#81] [Q&A] [answered]
I tried to follow the instructions for overriding library versions here: https://srid.ca/haskell-template/dependency

However, I am getting the following error:
```
error: attribute 'callCabal2Nix' missing

       at /nix/store/pf28nmzgqjy7sarx5rlhay83jharcrm0-source/flake.nix:30:34:

           29|           overrides = self: super: with pkgs.haskell.lib; {
           30|             ihp-hsx = dontCheck (self.callCabal2Nix "ihp-hsx" inputs.ihp-hsx {});
             |                                  ^
           31|           };
       Did you mean callCabal2nix?
(use '--show-trace' to show detailed location information)
```

**Comment by daanrs** [comment, marked as answer]
I am not the library author, so I might be wrong, but I will try my best to help! 

This error is luckily just a typo (I've made the same mistake before), if you use 
```nix
overrides = self: super: with pkgs.haskell.lib; {
    ihp-hsx = dontCheck (self.callCabal2nix "ihp-hsx"  inputs.ihp-hsx {});
};
```
it should work. 

Note that in this case you are overriding the ihp-hsx package with a flake input. If you for instance want to override the pre-existing package in nixpkgs to just disable the tests, then you can use something like
```nix
overrides = self: super: with pkgs.haskell.lib; {
    ihp-hsx = dontCheck super.ihp-hsx;
};
```

**Reply by billksun** [reply]
@daanrs, sharp eyes!

Yes, I want to override the `ihp-hsx` package with a version from the master branch in GitHub. After correcting for the typo (lower case "nix" in `callCabal2nix`, for those that are following), I got the "infinite recursion encountered error":
```
error: infinite recursion encountered

       at /nix/store/1zzy0v6laha6r67l0n3wqkgqiwvvj7hv-source/pkgs/stdenv/generic/make-derivation.nix:321:7:

          320|       depsHostHostPropagated      = lib.elemAt (lib.elemAt propagatedDependencies 1) 0;
          321|       propagatedBuildInputs       = lib.elemAt (lib.elemAt propagatedDependencies 1) 1;
             |       ^
          322|       depsTargetTargetPropagated  = lib.elemAt (lib.elemAt propagatedDependencies 2) 0;
(use '--show-trace' to show detailed location information)
```

Any ideas what might this mean? I originally was using `haskell.nix` with my project, and overriding the Hackage package with the Github version was as easy as adding the repo information in the `cabal.project` file; and it worked without problems.

**Reply by daanrs** [reply]
One thing that might cause that is if the flake input selects the wrong cabal file. I took a quick look at the ihp repo, and saw that the ihp-hsx package is defined inside of a directory, so the flake input uri would need to be
```nix
ihp-hsx.url = github:digitallyinduced/ihp?dir=ihp-hsx
```
Otherwise it might try to build the ihp package, for which ihp-isx is a dependency, so it might get into a infinite recursion error. If that is not the problem, then I'm not sure what is.

**Reply by billksun** [reply]
I think I have the subdir setup correctly. Here's a minimal reproducible repo: 
https://github.com/billksun/Nix-Haskell-Template-IHP-HSX-Infinite-Recursion-Min-Example

**Reply by srid** [reply]
I don't see infinite recursion with this, for some reason:

```diff
diff --git a/flake.nix b/flake.nix
index 1cdb4df..8353950 100644
--- a/flake.nix
+++ b/flake.nix
@@ -12,7 +12,7 @@
       # owner = "digitallyinduced";
       # repo = "ihp";
       # dir = "ihp-hsx";
-      url = github:digitallyinduced/ihp?dir=ihp-hsx;
+      url = "github:digitallyinduced/ihp";
       flake = false;
     };
   };
@@ -31,7 +31,7 @@
         # has only one.
         haskellProjects.main = {
           overrides = self: super: with pkgs.haskell.lib; {
-            ihp-hsx = dontCheck (self.callCabal2nix "ihp-hsx" inputs.ihp-hsx {});
+            ihp-hsx = dontCheck (self.callCabal2nix "ihp-hsx" (inputs.ihp-hsx + /ihp-hsx) {});
           };
           haskellPackages = pkgs.haskell.packages.ghc925;
           packages = {
```

**Reply by daanrs** [reply]
I guess selecting a directory in the input url doesn't work with non-flake inputs?

**Reply by billksun** [reply]
@srid, that indeed did the trick! Is this an issue with flakes or nix then?
